doc/source/development/window.rst

```diff
@@ -0,0 +1,187 @@
+Generalized Window Operations
+=============================
+
+Rationale
+---------
+The current window implementation in :file:`window.pyx` is difficult to
+customize and couples the creation of window bounds, loop over the values to
+aggregate, the logic for updating an aggregation.
+
+Requirements
+------------
+#. Pass all existing user-facing API tests
+#. User-defined aggregations performance on par with "builtins"
+   #. UDAs can be implemented the same way as builtins
+#. Ideally no performance regressions
+#. Performant support for custom offsets (e.g., custom business
+   days)
+#. New features are strict additions to the API
+
+Soft Requirements
+-----------------
+#. Lower the maintenance burden of the existing implementation
+#. Reduce code duplication
+
+Implementation
+--------------
+Here we describe details of the implementation that we think satisfy the
+requirements.
+
+Aggregation
+~~~~~~~~~~~
+We propose to split aggregations into two parts: the algorithm used to compute
+the value of an aggregation for a given row, and a class used for aggregation.
+
+The aggregation algorithm is an interface that when given a start and stop
+point representing indices into a :attr:`values` ndarray, knows how to
+efficiently compute the value of an aggregation in between start and stop.
+
+For example:
+
+.. code-block:: python
+
+   import abc
+   from typing import Union, Callable
+
+   class AggregationAlgorithm(abc.ABC):
+       def __init__(
+           self, values: ndarray, agg: Union[Callable[...], Agg]
+       ) -> None:
+           self.values = values
+           self.agg = agg
+
+       @abc.abstractmethod
+       def query(
+           self, start: int, stop: int, previous_start: int, previous_end: int
+       ) -> Output:
+           ...
+
+   class SegmentTreeAggregator(AggregationAlgorithm):
+       def query(
+           self, start: int, stop: int, previous_start: int, previous_end: int
+       ) -> Output:
+           ...
+
+   class SubtractableAggregator(AggregationAlgorithm):
+       def query(
+           self, start: int, stop: int, previous_start: int, previous_end: int
+       ) -> Output:
+           ...
+
+   class ApplyAggregator(AggregationAlgorithm):
+       def query(
+           self, start: int, stop: int, previous_start: int, previous_end: int
+       ) -> Output:
+           return self.agg(self.values[start:stop])
+
+
+An aggregation is an interface that implements methods to update its internal
+state and a :meth:`finalize` method to return the current value of the
+aggregation based on the internal state.
+
+An example implementation of a :class:`Sum` kernel would look similar to
+
+.. code-block:: python
+
+   from typing import Generic, TypeVar
+
+   Input1 = TypeVar("Input1")
+   Input2 = TypeVar("Input2")
+   Output = TypeVar("Output")
+
+   class UnaryAggregation(Generic[Input1, Output]):
```

Comment:
I think that's actually a bit overkill for the first iteration of the spec but is definitely something to consider in the implementation. In any event, only `finalize` could be in such a base class because `step` has a different number of arguments depending on the kind of kernel we're making (see my `Covariance` example).

Reply:
right I think the signature can be use general then. ``step`` is required IIRC, and ``finalize``is optional? the base class would make this clear.

Reply:
Yeah, I think this is too detailed for the spec. I would definitely agree that we want explicit protocols defined by ABCs, as opposed to wishy-washy protocols defined in obscure docstrings.